File: 6.迭代器模式/geqiangbao_20201125_iterator_OC/geqiangbao_20201125_iterator_OC/ViewController.m
//
//  ViewController.m
//  geqiangbao_20201125_iterator_OC
//
//  Created by xiaobao on 2020/11/25.
//

#import "ViewController.h"
#import "ArrayIterator.h"
#import "SetIterator.h"


@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self demo3];

    // Do any additional setup after loading the view.
}

-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    [self demo3];
}

// 系统迭代器
-(void)demo3{
//    正序迭代器
    NSArray *arr = @[@"1",@"2",@"33",@"444"];
    NSEnumerator<NSString*> *iterator = [arr objectEnumerator];
    NSString *str = @"";
    while ((str = iterator.nextObject)) {
        NSLog(@"======== %@",str);
    }
//    逆序迭代器
    NSEnumerator<NSString*> *iterator2 = [arr reverseObjectEnumerator];
    while ((str = iterator2.nextObject)) {
        NSLog(@"--------- %@",str);
    }
    
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:@"xiaoming" forKey:@"name"];
    [dict setObject:@"12" forKey:@"age"];
    NSEnumerator<id> *iterator3 = [dict.allKeys objectEnumerator];
    id result = NULL;
    while ((result = iterator3.nextObject)) {
        NSLog(@"%@",result);
    }
    NSEnumerator<id> *iterator4 = [dict.allValues objectEnumerator];
    while ((result = iterator4.nextObject)) {
        NSLog(@"%@",result);
    }
}


-(void)demo2{
    NSArray *arr = @[@"1",@"2",@"33",@"444"];
    NSSet *set = [NSSet setWithArray:arr];
    id<IteratorProtocol> iterator = [[SetIterator alloc]initWith:set];
    while (iterator.hasNext) {
        NSLog(@"%@",iterator.next);
    }
    [arr objectEnumerator];
}


/// 自定义迭代器
-(void)demo1{
    NSArray *arr = @[@"1",@"22",@"345"];
    id<IteratorProtocol> iterator = [[ArrayIterator alloc] initWith:arr];
    while (iterator.hasNext) {
        NSLog(@"%@",iterator.next);
    }
    
}


@end

File: 18.代理模式/geqiangbao_20201203_delegate_swift/geqiangbao_20201203_delegate_swift/ViewController.swift
//
//  ViewController.swift
//  geqiangbao_20201203_delegate_swift
//
//  Created by xiaobao on 2020/12/3.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        demo1()
        // Do any additional setup after loading the view.
    }

//    静态代理
    func demo1(){
        let proxy = PersonProxy(name: "小明")
        proxy.buyProduct()
        proxy.payGoods()
    }
    /*
     1.可以做到在不修改目标对象的功能前提下,对目标功能扩展.
     静态代理需要目标对象和代理对象实现相同的接口。可以在不修改目标对象功能的前提下.对目标功能进行扩展。


     2.缺点:

     因为代理对象需要与目标对象实现一样的接口,所以会有很多代理类,类太多
     同时,一旦接口增加方法,目标对象与代理对象都要维护.
     如何解决静态代理中的缺点呢?答案是可以使用动态代理方式
     */
   


}

